### PR TITLE
Traceroute SNR cleanup

### PR DESCRIPTION
**Pull request by bjpetit**
## What changed?

This change updates traceroute handling of SNR values.

1. For direct traceroute results, use the SNR from the routing message, rather than pulling it from the node entry. Pulling from the node entry causes every direct traceroute entry to show the current SNR for the node, rather than the SNR at the time of the traceroute
2. cleans up a couple spots where the SNR is divided by 4. The change casts them to Floats first then divides so that the decimals aren't truncated.
3. Clean up some incorrect snr checks against 0. SNR can be negative. An unknown SNR is represented as -32. When the SNR value is encoded as an int in the routing message, the value is -128.
4. Fix up some formatting issues with the routeString and routeBackString

## Why did it change?

These changes clean up some areas which were returning incorrect SNR values as well as some incorrect formatting of the route strings

## How is this tested?

This change was tested with multiple traceroute runs and results

## Checklist

- [x] My code adheres to the project's coding and style guidelines.
- [x] I have conducted a self-review of my code.
- [x] I have commented my code, particularly in complex areas.
- [ ] I have made corresponding changes to the documentation.
- [x] I have tested the change to ensure that it works as intended.

### CODE REVIEW

**Comment by bjpetit** [comment]
Here's an example of the traceroute output.
IMG_0526

File: Meshtastic/Helpers/BLEManager.swift [hunk 1]
```diff
@@ -848,7 +848,7 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 						connectedHop.time = Date()
 						connectedHop.num = connectedPeripheral.num
 						connectedHop.name = connectedNode.user?.longName ?? "???"
-						connectedHop.snr = Float(routingMessage.snrBack.last ?? 0 / 4)
+						connectedHop.snr = Float(routingMessage.snrBack.last ?? 0) / 4
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: Can we add code comments about why we are dividing by 4? (if anyone knows the reason)

File: Meshtastic/Helpers/BLEManager.swift [hunk 2]
```diff
@@ -866,7 +866,10 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 							let traceRouteHop = TraceRouteHopEntity(context: context)
 							traceRouteHop.time = Date()
 							if routingMessage.snrTowards.count >= index + 1 {
-								traceRouteHop.snr = Float(routingMessage.snrTowards[index] / 4)
+								traceRouteHop.snr = Float(routingMessage.snrTowards[index]) / 4
+							} else {
+								// If no snr in route, try last snr from node
+								traceRouteHop.snr = hopNode?.snr ?? 0
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: Why we are pulling readings from the nodeDb here?

Isn't this going to possibly be `SNR(selfNode <- currentHopNode)` when we actually want, `SNR(currentHopNode <- previousHopNode)` for the path of the trace? (the only time this would maybe be correct is if we are certain the nodeDB updated from the trace message already and this was the final hop)

I'd much rather see us display "unknown" if trace data is missing than something we know is likely incorrect.

File: Meshtastic/Helpers/BLEManager.swift [hunk 3]
```diff
@@ -835,8 +835,8 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 					let traceRoute = getTraceRoute(id: Int64(decodedInfo.packet.decoded.requestID), context: context)
 					traceRoute?.response = true
 					if routingMessage.route.count == 0 {
-						let snr = routingMessage.snrBack.count > 0 ? routingMessage.snrBack[0] / 4 : 0
-						traceRoute?.snr = Float(snr)
+						let snr = routingMessage.snrBack.count > 0 ? Float(routingMessage.snrBack[0]) / 4 : 0.0
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: zero is a bad default value for these unless it's impossible for SNR reading values to have that. It seems like it's possible tho since it frequently ranges from -20 to +10.

File: Meshtastic/Helpers/BLEManager.swift [hunk 4]
```diff
@@ -898,8 +901,8 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 						}
 						hopNodes.append(destinationHop)
 						/// Add the destination node to the end of the route towards string and the beginning of teh route back string
-						routeString += "\(traceRoute?.node?.user?.longName ?? "unknown".localized) \((traceRoute?.node?.num ?? 0).toHex()) \(traceRoute?.node?.snr ?? 0 > 0 ? traceRoute?.node?.snr ?? 0 : 0.0)dB)"
-						var routeBackString = "\(traceRoute?.node?.user?.longName ?? "unknown".localized) \((traceRoute?.node?.num ?? 0).toHex()) \(traceRoute?.node?.snr ?? 0 > 0 ? traceRoute?.node?.snr ?? 0 : 0.0)dB) --> "
+						routeString += "\(traceRoute?.node?.user?.longName ?? "unknown".localized) \((traceRoute?.node?.num ?? 0).toHex()) (\(traceRoute?.node?.snr ?? traceRoute?.node?.snr ?? 0.0)dB)"
+						var routeBackString = "\(traceRoute?.node?.user?.longName ?? "unknown".localized) \((traceRoute?.node?.num ?? 0).toHex()) (\(traceRoute?.node?.snr ?? traceRoute?.node?.snr ?? 0.0)dB) --> "
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: Thoughts about using a default of something like `"??"` (unknown) rather than `0.0`? 

Default 0.0 is confusing since it's in the range of possible values.

File: Meshtastic/Helpers/BLEManager.swift (hunk 1 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: I'll put a comment in this code. 
The 'quick' answer is that for the routing messages the field is an int32, scaled buy 4. This then represents a float snr as an int. You can see this noted in the protobuf spec for the Routing messages.

File: Meshtastic/Helpers/BLEManager.swift (hunk 3 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Agreed. Let me rework these spots a bit.

File: Meshtastic/Helpers/BLEManager.swift (hunk 1 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: I've updated the comments on what we are doing with the snr fields from the routing message. I've also noted the snr value that represents unknown.

File: Meshtastic/Helpers/BLEManager.swift (hunk 2 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: I've removed the use of the node snr here. If the routing message doesn't have a valid snr we make it unknown to reflect what really happened during the traceroute.

File: Meshtastic/Helpers/BLEManager.swift (hunk 3 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: The firmware will set unknown/unset snr values in routing messages to -128. When we divide by 4 that becomes -32 which is outside the range of valid snr values.
This code is now setting the snr as unknown when it sees -32.

File: Meshtastic/Helpers/BLEManager.swift (hunk 4 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Done.

File: Meshtastic/Helpers/BLEManager.swift [hunk 5]
```diff
@@ -898,8 +906,8 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 						}
 						hopNodes.append(destinationHop)
 						/// Add the destination node to the end of the route towards string and the beginning of teh route back string
```
Comment (GUVWAF) [inline Meshtastic/Helpers/BLEManager.swift]: In general this looks good to me. 

One thing is that if `hop_start` is not set in the traceroute response (due to nodes with firmware <2.3 in the route), `routeBack` might be incomplete, since we can't add unknown nodes and SNR in this case (because we don't know how many missing hops there were). I think it would be best to skip showing the route back in this case, like what is done for [Android](https://github.com/meshtastic/Meshtastic-Android/blob/b668a21b68465c13d15ca224df785c86ebe4f3ad/app/src/main/java/com/geeksville/mesh/model/RouteDiscovery.kt#L17) and the [CLI](https://github.com/meshtastic/python/blob/646aa981d5fc191f1900da830615dec91cc77403/meshtastic/mesh_interface.py#L584).

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: I've put a conditional around the routeBack logic. This should match the logic seen in the Andriod app.

File: Meshtastic/Helpers/BLEManager.swift [hunk 6]
```diff
@@ -883,12 +890,13 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 								}
 							}
 							hopNodes.append(traceRouteHop)
-							routeString += "\(hopNode?.user?.longName ?? (node == 4294967295 ? "Repeater" : String(hopNode?.num.toHex() ?? "unknown".localized))) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr > 0 ? hopNode?.snr ?? 0.0 : 0.0)dB) --> "
+							routeString += "\(hopNode?.user?.longName ?? (node == 4294967295 ? "Repeater" : String(hopNode?.num.toHex() ?? "unknown".localized))) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized)dB) --> "
```
Comment (GUVWAF) [inline Meshtastic/Helpers/BLEManager.swift]: One more thing: "Repeater" is not really accurate, it can also just be a node that doesn't have the encryption key, or for the route back it can be old firmware.

It's used in multiple places. I would just also use "unknown" here.

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: In that case, would it be reasonable to just drop the "node == 4294967295" handling and let it fall through to the remaining options? It would try to use the node number or fall back to unknown.

Comment (GUVWAF) [reply Meshtastic/Helpers/BLEManager.swift]: Yes, sounds reasonable to me.

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Done.

**Review by GUVWAF — APPROVED** [review]

File: Meshtastic/Helpers/BLEManager.swift [hunk 7]
```diff
@@ -835,8 +835,11 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 					let traceRoute = getTraceRoute(id: Int64(decodedInfo.packet.decoded.requestID), context: context)
 					traceRoute?.response = true
 					if routingMessage.route.count == 0 {
-						let snr = routingMessage.snrBack.count > 0 ? routingMessage.snrBack[0] / 4 : 0
-						traceRoute?.snr = Float(snr)
+						// Routing messages snr values are snr * 4 stored as an int
+						// If a traceroute snr value is unknown this field will contain INT8_MIN or -128
+						// After converting to a float here, -32 is our unknown value.
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift:840]: 🥇

File: Meshtastic/Helpers/BLEManager.swift [hunk 8]
```diff
@@ -883,12 +890,13 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 								}
 							}
 							hopNodes.append(traceRouteHop)
-							routeString += "\(hopNode?.user?.longName ?? (node == 4294967295 ? "Repeater" : String(hopNode?.num.toHex() ?? "unknown".localized))) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr > 0 ? hopNode?.snr ?? 0.0 : 0.0)dB) --> "
+							routeString += "\(hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized)dB) --> "
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: nit: this is kinda hard to read with text wrapping.
```
routeString += "\(hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized)dB) --> "
```

```
let hopName = hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)
let mqttLabel = hopNode?.viaMqtt ?? false ? "MQTT" : ""
let snrLabel = (traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized

routeString += "\(hopName) \(mqttLabel) (\snrLabel)dB) --> "
```
^ after refactoring it becomes clear that we have an unnecessary whitespace if mqttLabel is blank.

```
let hopName = hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)
let mqttLabel = hopNode?.viaMqtt ?? false ? "MQTT " : ""
let snrLabel = (traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized

routeString += "\(hopName) \(mqttLabel)(\snrLabel)dB) --> "
```
^ fixt `"MQTT "` + `\(hopName) \(mqttLabel)(\snrLabel)`

File: Meshtastic/Helpers/BLEManager.swift [hunk 9]
```diff
@@ -897,41 +905,50 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
...
+									}
+								}
+								traceRouteHop.num = hopNode?.num ?? 0
+								if hopNode != nil {
+									if decodedInfo.packet.rxTime > 0 {
+										hopNode?.lastHeard = Date(timeIntervalSince1970: TimeInterval(Int64(decodedInfo.packet.rxTime)))
+									}
+								}
+								hopNodes.append(traceRouteHop)
+								routeBackString += "\(hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized)dB) --> "
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: same nit about formatting + whitespace for mqtt string

File: Meshtastic/Helpers/BLEManager.swift (hunk 8 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Yes. That's more readable. Will make that update.

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Done

File: Meshtastic/Helpers/BLEManager.swift (hunk 9 above)
Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Also done.

File: Meshtastic/Helpers/BLEManager.swift [hunk 10]
```diff
@@ -942,7 +946,11 @@ class BLEManager: NSObject, CBPeripheralDelegate, MqttClientProxyManagerDelegate
 									}
 								}
 								hopNodes.append(traceRouteHop)
-								routeBackString += "\(hopNode?.user?.longName ?? (node == 4294967295 ? "Repeater" : String(hopNode?.num.toHex() ?? "unknown".localized))) \(hopNode?.viaMqtt ?? false ? "MQTT" : "") (\(traceRouteHop.snr != -32 ? String(traceRouteHop.snr) : "unknown ".localized)dB) --> "
+
+								let hopName = hopNode?.user?.longName ?? String(hopNode?.num.toHex() ?? "unknown".localized)
```
Comment (powersjcb) [inline Meshtastic/Helpers/BLEManager.swift]: `node == 4294967295` was previously hardcoded here (value of 2^32-1). Any idea why that was the case?

edit: oh, just kidding part of the removed logic for identifying repeaters.

lgtm

**Review by powersjcb — APPROVED** [review]

File: Meshtastic/Helpers/BLEManager.swift (hunk 10 above)
Comment (garthvh) [reply Meshtastic/Helpers/BLEManager.swift]: The logic identifying routers needs to be in there, why did it get removed?

Comment (GUVWAF) [reply Meshtastic/Helpers/BLEManager.swift]: It's not necessarily a router (or repeater). It can be a node that can't decrypt because it doesn't have the channel key, or in the route back it's likely a node on firmware before 2.5.

The only case that could cause a repeater to show as `0xFFFFFFFF` as well is if it was set to `ALL_SKIP_DECODING`.

Comment (garthvh) [reply Meshtastic/Helpers/BLEManager.swift]: Could say repeater (no key) or something but it needs to be in there.  Most users are just using the app and are not digging through the docs.

Comment (GUVWAF) [reply Meshtastic/Helpers/BLEManager.swift]: Why not just "unknown"? There are multiple options (old firmware is another), and we don't which one it is.

If it's a repeater not set to `ALL_SKIP_DECODING`, it will be its NodeNum, not `0xFFFFFFFF`.

Comment (GUVWAF) [reply Meshtastic/Helpers/BLEManager.swift]: Ah, or will it show `0xFFFFFFFF` now when it doesn't know the long name? That would indeed be confusing.

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Thoughts on next steps here?
It sounds like adding back special handling for 0xFFFFFFFF makes sense. But should it call that out as 'unknown' since a hop could come back with that set for multiple reasons? Or is something more precise that these entries should be named?

Comment (powersjcb) [reply Meshtastic/Helpers/BLEManager.swift]: `unknown` feels like a clean answer if we cannot say who they are for any reason

Comment (garthvh) [reply Meshtastic/Helpers/BLEManager.swift]: > Thoughts on next steps here? It sounds like adding back special handling for 0xFFFFFFFF makes sense. But should it call that out as 'unknown' since a hop could come back with that set for multiple reasons? Or is something more precise that these entries should be named?

Yes I want the special handling in there

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: Should the hopName end up as "unknown" or something else in this path?

Comment (garthvh) [reply Meshtastic/Helpers/BLEManager.swift]: > Should the hopName end up as "unknown" or something else in this path?

Repeater like it was.

Comment (bjpetit) [reply Meshtastic/Helpers/BLEManager.swift]: done

**Review by garthvh — APPROVED** [review]
Looks good
